<?xml version="1.0"?>

<robot name="traffic_sign" xmlns:xacro="http://ros.org/wiki/xacro">

    <!-- Constants for robot dimensions -->
    <xacro:property name="radius" value="0.02" />
    <xacro:property name="length" value="0.25" />
    <xacro:property name="damping" value="0.1" />
    <xacro:property name="mass" value="0.5" />
    <xacro:property name="effort_limit" value="10" />
    <xacro:property name="velocity_limit" value="100" />


    <material name="orange">
        <color rgba="${255/255} ${108/255} ${10/255} 1.0" />
    </material>

    <material name="red">
        <color rgba="0.8 0.0 0.0 1.0" />
    </material>

    <material name="black">
        <color rgba="0.0 0.0 0.0 1.0" />
    </material>

    <!-- World -->
    <link name='world' />

    <joint name="joint_base" type="fixed">
        <!-- <origin xyz="27.86 11.9 0.0" rpy= "0 0 -1.566" /> -->
        <origin xyz="0.0 0.0 0.0" rpy="0 0 -1.566" />
        <parent link="world" />
        <child link="stand_link" />
    </joint>

    <!-- Stand link -->
    <link name='stand_link'>
        <inertial>
            <origin xyz="-0.00 0.0 0.5" rpy="0 0 0" />
            <inertia ixx="1.11" ixy="0.0" ixz="0.0" iyy="100.11" iyz="0.0" izz="1.01" />
            <mass value="101.0" />
        </inertial>
        <!-- Base -->
        <visual>
            <origin xyz="0 0.0 0.01" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.2" length="0.02" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <origin xyz="0 0.0 0.01" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.2" length="0.02" />
            </geometry>
        </collision>
        <!-- Pole -->
        <visual>
            <origin xyz="0 0.0 0.5" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.02" length="1.0" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <origin xyz="0 0.0 0.5" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.02" length="1.0" />
            </geometry>
        </collision>
        <!-- Cross bar -->
        <visual>
            <origin xyz="0 0.0 1.0" rpy="0 -1.57 0" />
            <geometry>
                <cylinder radius="0.02" length="0.8" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <origin xyz="0 0.0 1.0" rpy="0 -1.57 0" />
            <geometry>
                <cylinder radius="0.02" length="0.8"/>
            </geometry>
        </collision>
    </link>


    <!-- red_link -->
    <link name='red_link'>
        <inertial>
            <origin xyz="-0.00 -0.00 ${length/2}" rpy="0 0 0" />
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0104" iyz="0.0" izz="0.0" />
            <mass value="${mass}" />
        </inertial>
        <!-- Base -->
        <visual>
            <origin xyz="0 0 ${length/2}" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${length}" />
            </geometry>
            <material name="red" />
        </visual>
        <collision>
            <origin xyz="0 0 ${length/2}" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${length}" />
            </geometry>
        </collision>
        <!-- Sign -->
        <visual>
            <origin xyz="0 -0.04 ${length/2}" rpy="1.57 0 0" />
            <geometry>
                <cylinder radius="0.2" length="0.01" />
            </geometry>
            <material name="red" />
        </visual>
        <collision>
            <origin xyz="0 -0.04 ${length/2}" rpy="1.57 0 0" />
            <geometry>
                <cylinder radius="0.2" length="0.01" />
            </geometry>
        </collision>
    </link>

    <joint name="joint_red" type="continuous">
        <origin xyz="-0.2 0 1.0" rpy="0 0 0" />
        <child link="red_link" />
        <parent link="stand_link" />
        <axis xyz="1 0 0" />
        <dynamics damping="${damping}" friction="0.0" />
        <limit effort="${effort_limit}" velocity="${velocity_limit}" />
    </joint>

    <!-- green link -->
    <link name='green_link'>
        <inertial>
            <origin xyz="-0.00 -0.00 0.125" rpy="0 0 0" />
            <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0104" iyz="0.0" izz="0.0" />
            <mass value="${mass}" />
        </inertial>
        <!-- Base -->
        <visual>
            <origin xyz="0 0 ${length/2}" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${length}" />
            </geometry>
            <material name="green" />
        </visual>
        <collision>
            <origin xyz="0 0 ${length/2}" rpy="0 0 0" />
            <geometry>
                <cylinder radius="${radius}" length="${length}" />
            </geometry>
        </collision>
        <!-- Sign -->
        <visual>
            <origin xyz="0 -0.04 ${length/2}" rpy="1.57 0 0" />
            <geometry>
                <cylinder radius="0.2" length="0.01" />
            </geometry>
            <material name="green" />
        </visual>
        <collision>
            <origin xyz="0 -0.04 ${length/2}" rpy="1.57 0 0" />
            <geometry>
                <cylinder radius="0.2" length="0.01" />
            </geometry>
        </collision>
    </link>

    <joint name="joint_green" type="continuous">
        <origin xyz="0.2 0 1.0" rpy="0 0 0" />
        <child link="green_link" />
        <parent link="stand_link" />
        <axis xyz="1 0 0" />
        <dynamics damping="${damping}" />
        <limit effort="${effort_limit}" velocity="${velocity_limit}" />
    </joint>


    <transmission name="tran1">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_red">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor1">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="tran2">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_green">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor2">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>


    <!-- ros_control plugin -->
    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/traffic_sign</robotNamespace>
            <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
        </plugin>
    </gazebo>

    <!-- base_link -->
    <gazebo reference="stand_link">
        <material>Gazebo/Black</material>
    </gazebo>

    <!-- link_1 -->
    <gazebo reference="red_link">
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
        <material>Gazebo/Red</material>
    </gazebo>

    <!-- link_2 -->
    <gazebo reference="green_link">
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
        <material>Gazebo/Green</material>
    </gazebo>

</robot>